speechbrain/utils/autocast.py

```diff
@@ -3,13 +3,131 @@
 
 Authors
  * Sylvain de Langen 2023
+ * Adel Moumen 2025
 """
 
 import functools
 from typing import Callable, Optional
+from dataclasses import dataclass
 
 import torch
+from contextlib import nullcontext
 
+@dataclass
+class AMPConfig:
+    """Configuration for automatic mixed precision (AMP).
+
+    Arguments
+    ---------
+    dtype : torch.dtype
+        The dtype to use for AMP.
+    """
+
+    dtype: torch.dtype
+
+    @classmethod
+    def from_name(self, name):
+        """Create an AMPConfig from a string name.
+
+        Arguments
+        ---------
+        name : str
+            The name of the AMPConfig to create.  Must be one of `fp32`,
+            `fp16`, or `bf16`.
+
+        Returns
+        -------
+        AMPConfig
+            The AMPConfig corresponding to the name.
+        """
+        if name is None or name == "fp32":
+            return AMPConfig(torch.float32)
+        elif name == "fp16":
+            return AMPConfig(torch.float16)
+        elif name == "bf16":
+            return AMPConfig(torch.bfloat16)
+        else:
+            raise ValueError(
+                f"Specified autocast mode ({name}) incorrect, expected one of `fp32`, `fp16`, `bf16`."
+            )
+        
+class TorchAutocast:
+    """
+    Utility for conditionally enabling torch.autocast.
+
+    This context manager activates autocasting (via ``torch.autocast``)
+    when the ``enabled`` flag is True. When disabled, it behaves as a no-op
+    context manager.
+
+    Arguments
+    ---------
+    enabled : bool
+        Whether to enable torch.autocast.
+    *args
+        Positional arguments to pass to torch.autocast.
+    **kwargs
+        Keyword arguments to pass to torch.autocast.
+
+    Examples
+    --------
+    >>> with TorchAutocast(enabled=True, device="cpu", dtype=torch.bfloat16):
+    ...     # Run model inference with autocasting enabled.
+    """
+
+    def __init__(self, enabled: bool, *args, **kwargs):
+        if enabled:
+            self.context = torch.autocast(*args, **kwargs)
+        else:
+            self.context = nullcontext()  # no-op context manager
```

Comment:
Just to make sure I understand what is happening here, the change here is to add the `enabled` flag, otherwise `torch.autocast` would work fine as-is, is that right?

Reply:
Yeah. If `enabled` is `False` then it is just a no-op context manager, meaning nothing is applied while if you have `enabled` to be `True` then it acts as a regular autocast ctx.